docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+    fn read(data: FontData<'a>) -> Result<Self, ReadError> {
+        let mut cursor = data.cursor();
+        let version: u16 = cursor.read()?;
+        cursor.advance::<u32>(); // some_val
+        cursor.advance::<u32>(); // other_val
+        let flags_count: u16 = cursor.read()?;
+        let flags_byte_len = flags_count as usize * u16::RAW_BYTE_LEN;
+        cursor.advance_by(flags_byte_len); // flags
+        let versioned_value_byte_start = version
+            .compatible(1)
+            .then(|| cursor.position())
+            .transpose()?;
+        version.compatible(1).then(|| cursor.advance::<u32>());
+        cursor.finish(FoobMarker {
+            flags_byte_len,
+            versioned_value_byte_start,
+        })
+    }
+}
+```
+
+Let's walk through this. Firstly, the whole process is based around a 'cursor'
+type, which is simply a way of advancing through the input data on a
+field-by-field basis. Where we need to know the value of a field in order to
+validate subsequent fields, we read that field into a local variable.
+Additionally, values that we have to compute based on other fields are currently
+cached in the marker struct, although this is an implementation detail and may
+change. Let's walk through this code, field by field:
+
+- **version**: as this is marked with the `#[version]` attribute, we read the
+  value into a local variable, since we will need to know the version when
+  reading any versioned fields.
+- **some_val**: this is a simple value, and we do not need to know what it is,
+  only that it exists. We `advance` the cursor by the appropriate number of
+  bytes.
+- **other_val**: ditto. The compiler will be able to combine these two
+  `advances` into a single operation.
+- **flags_count**: This value is referenced in the `#[count]` attribute on the
+  following field, and so we bind it to a local variable.
+- **flags**: the `#[count]` attribute indicates that the length of this array is
+  stored in the `flags_count` field. We determine the array length by
+  multiplying that value by the size of the array member, and we advance the
+  cursor by that number of bytes.
+- **versioned_value**: this field is only available if the `version` field is >=
+  to `1` (this is specified via the `#[available]` attribute). We record the
+  current cursor position (as an `Option`, which will be `Some` only if the
+  version is compatible) and then we advance the cursor by the size of the
+  field's type.
+
+Finally, having finished with each field, we call the `finish` method on the
+cursor: this performs a final bounds check, and instantiates the table with the
+provided marker.
+
+> ***note***:
+>
+> The `FontRead` trait is currently doing a bit of a double duty: in the case of
+> tables, it is expected to perform a very minimal validation (essentially just
+> bounds checking) but in the case of records it serves as an actual parse
+> function, returning a concrete instance of the type. It is possible that these
```

Comment:
suggest explaining why

Reply:
The point of these blockquote notes is to point out things that I think are a bit awkward, e.g. I really don't have a good explanation of why 😒